From: John Fleck <john.fleck@intel.com>
Date: Sat, 12 May 2018 14:03:25 -0400
Subject: Fix for nodes dropping off from IPoIB mcast group

Check if the old topology port was DOWN when determining if
ClientReregister should be sent to handle the case where the
FM marked a port DOWN on an error path.

This is needed because if the port was marked DOWN, then we
will have cleaned up the MC groups.

Update version in spec file as well as GitHub url

Signed-off-by: John Fleck <john.fleck@intel.com>
Signed-off-by: Michael Brooks <michael.brooks@intel.com>
---
 Esm/build.env                   |  2 +-
 Esm/ib/src/smi/sm/sm_activate.c | 24 ++++++++++++++++++------
 2 files changed, 19 insertions(+), 7 deletions(-)

diff --git a/Esm/build.env b/Esm/build.env
index 33c76fd..d86a259 100644
--- a/Esm/build.env
+++ b/Esm/build.env
@@ -1,5 +1,5 @@
 # Adjust the environment variables if necessary
 export PRODUCT=VIEO_HOST
-export RELEASE_TAG=10_7_0_0_141
+export RELEASE_TAG=10_7_0_0_142
 export BUILD_CONFIG=${BUILD_CONFIG:-"release"}
 export BUILD_WITH_STACK=OPENIB
diff --git a/Esm/ib/src/smi/sm/sm_activate.c b/Esm/ib/src/smi/sm/sm_activate.c
index 2f6d5a3..0201e0b 100644
--- a/Esm/ib/src/smi/sm/sm_activate.c
+++ b/Esm/ib/src/smi/sm/sm_activate.c
@@ -34,14 +34,26 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 static __inline__ int
 needs_reregistration(Node_t * nodep, Port_t * portp, pActivationRetry_t retry)
 {
-	if (activation_retry_attempts(retry)) {
-		// on retries, only reregister if pending
+	// general conditions are only checked on first attempt.
+	// on retries, only reregister if it remains pending from previous attempts
+	if (activation_retry_attempts(retry))
 		return portp->portData->reregisterPending;
-	}
 
-	return topology_passcount < 1 // new sm
-		|| sm_find_guid(&old_topology, nodep->nodeInfo.NodeGUID) == NULL // new node
-		|| portp->portData->reregisterPending; // previous attempts haven't yet succeeded
+	// reregister if new sm
+	if (topology_passcount < 1) return TRUE;
+
+	// reregister if previous attempts haven't yet succeeded
+	if (portp->portData->reregisterPending) return TRUE;
+
+	// reregister if node/port is new this sweep
+	Node_t * oldnodep = sm_find_guid(&old_topology, nodep->nodeInfo.NodeGUID);
+	if (!oldnodep) return TRUE;
+	Port_t * oldportp = sm_find_node_port(&old_topology, oldnodep, portp->index);
+	if (!oldportp) return TRUE;
+
+	// reregister if old port was DOWN
+	// (implying we removed its groups/subscriptions)
+	return oldportp->state <= IB_PORT_DOWN;
 }
 
 static void
